### ISSUE TITLE
[BUG] Azure Batch unable to submit jobs greater than a certain amount of tasks; exception thrown isn't clear

### ISSUE THREAD

**Issue by steveellis** [Azure/azure-sdk-for-net#8764]
**Description**
If we submit a job that is too large we are expecting to get a message that quotas are exceeded or similar. Instead, the .NET client library throws a generic Exception (not a BatchException) which seems to have little connection to the underlying issue. If we submit the job in smaller chunks (fewer tasks), the job runs fine. Is this behavior to be expected? Is it documented somewhere? How should we work within the limits?

**Stack Trace**
This is the exception that is thrown. It would appear that the job is being marked as completed before it has a chance to start, probably because it takes too long to add the tasks and times out. Adding a longer timeout to the AddTasksAsync call has no effect.

Here's a screenshot of this exception in VS:

![Snip20191113_55](https://user-images.githubusercontent.com/104455/68804234-45a6a480-062f-11ea-90d9-0c29a4537462.png)

**To Reproduce**
In our case we believe the quota we may be exceeding is the vCPU quota. For example our vCPU quota is 800. If we submit a job with more tasks than that quota, even though it doesn't exceeed our job or task quota or pool quota, the job fails in the above manner.

Our job quota is 1000, our pool quota is 100. **If we submit a job with ~> 1000 tasks we get the above behavior.** Neither pools nor jobs are exceeded in our case though.

This is all assuming it has something to do with quotas, which isn't at all clear. But generally to reproduce you would probably just need to submit a "large" job. I'm describing what that appears to be in our case. 

**Code Snippet**
This is how we're adding tasks to the job.
```
        protected async Task AddTasksToJob(string jobId, List<CloudTask> tasks)
        {
            try
            {
                await _batchClient.JobOperations.AddTaskAsync(jobId, tasks); // Would some of the other parameters available here help? Unclear from docs. Adding a timeout param has no effect.
            }
            catch (BatchException e) // <-- this doesn't catch because it's not a BatchException
            {
                if (e.RequestInformation?.BatchError?.Code == BatchErrorCodeStrings.PoolQuotaReached)                 
                {
                    Console.WriteLine("PoolQuota reached"); // Would expect something like this to happen but it doesn't.
                }
                throw e;
            }
        }
```

**Expected behavior**
Would expect the job to run, allocating the vCPUs or other resources as they become available when other tasks complete.

**Job Characteristics**
VM SKU: Standard_F8s_v2 (this has 8 vcpus)
Low priority nodes: 2
Dedicated nodes: 2

**Setup**
 Microsoft.Azure.Batch version: **12.0.0**

**Comment by matthchr** [comment]
Hey Steve - the exception that was thrown was `Microsoft.Azure.Batch.ParallelOperationsException` - which inherits from `AggregateException` and not `BatchException`. I agree that this is somewhat confusing and I've made a note to discuss internally what we can do about it. It's possibly worth noting that _inside_ the `ParallelOperationsException` is a collection of `BatchException`.

In any case the issue as per your shared exception screenshot is that: "The job is being terminated" - it looks like you specify `OnAllTasksComplete` when you submit the job, is that correct?

We have [some documentation](https://docs.microsoft.com/en-us/rest/api/batchservice/job/add?redirectedfrom=MSDN#request-body) which outlines what to do in this somewhat tricky case (see the OnAllTasksComplete entry), reproduced here:
> The action the Batch service should take when all Tasks in the Job are in the completed state.
Note that if a Job contains no Tasks, then all Tasks are considered complete. This option is therefore most commonly used with a Job Manager task; if you want to use automatic Job termination without a Job Manager, you should initially set onAllTasksComplete to noaction and update the Job properties to set onAllTasksComplete to terminatejob once you have finished adding Tasks. The default is noaction.

I think the problem (as mentioned above) is that since your job initially has no tasks in it, it immediately gets sent to `terminating` state. Then when you're trying to add tasks to it, the adds fail (you're racing with our termination though, so sometimes the adds make it in time, possibly with a higher chance to make it on time when adding fewer tasks as there's less initial processing/serialization overhead before the requests actually make it to the server).

**Comment by steveellis** [comment]
Hi Matt. Thanks for the quick reply and for noticing the type of the exception returned. If I had noticed that I might have seen the collection of  ```BatchException``` objects which could be inspected for the error code like I was expecting--which was a little off the mark. I was thinking it had something to do with quotas but instead is a misunderstanding of how adding tasks to a job performs.

You're correct that we're setting ```OnAllTasksComplete``` to ```TerminateJob``` when the job is created. This this how we do it:
```
CloudJob job = CreateJobWithAutoPool(); // Our method.
job.OnAllTasksComplete = OnAllTasksComplete.TerminateJob; // Set the property before committing.
await job.CommitAsync();
var tasks = new List<CloudTask>(); // Create the tasks.
// Snip. We add our tasks to the list. And then do this:
await _batchClient.JobOperations.AddTaskAsync(jobId, tasks);
```
That's enough to get it going... But as you say it seems like it's in a race to get the tasks into the job before the job itself times out.

The doc you quoted suggests to first set the job to ```OnAllTasksComplete.NoAction``` and then update the job after adding the tasks to ```OnAllTasksComplete.TerminateJob```. This seems sort of ok, although not exactly intuitive. I guess I'd expect that it would just keep the job around and wait for tasks. Because...why not? But I see the challenge of deciding how long to wait. Maybe reset the timer each time a task arrives or just set the time out to be a day or two instead of what seems to be a few mins. Or why not let my job max time set the lifetime. I'm sure there's more to it than that though... Just some thoughts.

Happy to do what is suggested but I can't see how to update a job property using the .NET client lib though. I can see the REST method for updating here: https://docs.microsoft.com/en-us/rest/api/batchservice/job/update

But I don't see an equivalent in JobOperations here: https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.batch.joboperations?view=azure-dotnet

How would I go about updating a job property with the .NET client lib after the job has been committed and tasks are being added? Or is some other pattern of creating jobs and tasks that would work better for us?

**Comment by steveellis** [comment]
Found method to update a job with the .NET SDK: it's ```CommitChanges`` on CloudJob. Was thrown because it wasn't in ```JobOperations```. Another thing: retrieve the job that you want to update first. Reusing the reference to the original doesn't work.

So after setting ```OnAllTasksComplete.NoAction``` when creating the job. Do this after the tasks have been added to the job. A little counterintuitive but it appears to work, allowing "large numbers" of tasks to a job.
```
var jobToUpdate = await _batchClient.JobOperations.GetJobAsync(job.Id);
jobToUpdate.OnAllTasksComplete = OnAllTasksComplete.TerminateJob;
await jobToUpdate.CommitChangesAsync();
```

**Comment by matthchr** [comment]
Hey Steve, sorry for not getting back before you figured it out.
Yes, `CommitChanges` will do a `PATCH` on the Job, alternatively you can also do `job.Commit` (after making the modifications on the job object) which will do an update (full-replace). Either one will work.

For this: 
> Another thing: retrieve the job that you want to update first. Reusing the reference to the original doesn't work.

`job.Refresh` will also allow you to then update + `CommitChanges`

As for as the awkwardness here - yes we definitely had discussions when we were designing this feature about what the right behavior in the case that there weren't ever any tasks in the job. 

I can give a bit of background:
As you highlighted, the biggest problem is that since jobs are not "closed" (i.e. you can add tasks to them even after you create them, possibly _long_ after you create them). In this "there are no tasks case" it's possible we could have come up with a good behavior (we give you a few minutes to add tasks before starting the auto-complete timers), but given the disparity of different customers workloads we didn't want to penalize customers with relatively short running jobs (say, 1-5m) with an extra overhead before autocomplete kicked in. Eventually we decided that even though it wasn't super intuitive, the cleanest thing from an intent perspective was just for the customer to explicitly tell us when they were "done" adding tasks (or at least, when they wanted us to start trying to autocomplete), which is how we landed on this "update the job when you're ready" case.

Glad to hear you've got things working now.

**Comment by steveellis** [comment]
Thanks Matt. No worries. I'm sure I'm not groking all of the timing challenges with a service like this. Good to know about the other ways to update the job.
